coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala

```diff
@@ -0,0 +1,150 @@
+
+    val tsdbQueryParams = qContext.origQueryParams
+
+    if(!tsdbQueryParams.isInstanceOf[PromQlQueryParams] || // We don't know the promql issued (unusual)
+      (tsdbQueryParams.isInstanceOf[PromQlQueryParams]
+        && !tsdbQueryParams.asInstanceOf[PromQlQueryParams].processRouting)) // Query was part of routing
+      localPlanner.materialize(logicalPlan, qContext)
+
+    else logicalPlan match {
+      case lp: BinaryJoin          => materializeBinaryJoin(lp, qContext)
+      case lp: LabelValues         => materializeLabelValues(lp, qContext)
+      case lp: SeriesKeysByFilters => materializeSeriesKeysFilters(lp, qContext)
+      case _                       => materializeSimpleQuery(logicalPlan, qContext)
+
+    }
+  }
+
+  private def getRoutingKeys(logicalPlan: LogicalPlan) = dataset.options.nonMetricShardColumns
+    .map(x => (x, LogicalPlan.getLabelValueFromLogicalPlan(logicalPlan, x)))
+
+  private def generateRemoteExec(queryContext: QueryContext, startMs: Long,
+                                 endMs: Long, endPoint: String): PromQlExec = {
+    val queryParams = queryContext.origQueryParams.asInstanceOf[PromQlQueryParams]
+    val promQlParams = PromQlQueryParams(ConfigFactory.parseString(s"""endpoint = "${endPoint}""""),
+      queryParams.promQl, startMs / 1000, queryParams.stepSecs, endMs / 1000, queryParams.spread, processFailure = true,
+      processRouting = false)
+    PromQlExec(queryContext, InProcessPlanDispatcher, dataset.ref, promQlParams)
+  }
+
+  def materializeSimpleQuery(logicalPlan: LogicalPlan, qContext: QueryContext): ExecPlan = {
+
+    val routingKeys = getRoutingKeys(logicalPlan)
+    if (routingKeys.forall(_._2.isEmpty)) localPlanner.materialize(logicalPlan, qContext)
+    else {
+      val routingKeyMap = routingKeys.map(x => (x._1, x._2.get.head)).toMap
+      val offsetMs = LogicalPlanUtils.getOffsetMillis(logicalPlan)
+      val periodicSeriesTime = getPeriodicSeriesTimeFromLogicalPlan(logicalPlan)
+      val periodicSeriesTimeWithOffset = TimeRange(periodicSeriesTime.startMs - offsetMs,
+        periodicSeriesTime.endMs - offsetMs)
+      val lookBackMs = getLookBackMillis(logicalPlan)
+      val stepMs = logicalPlan.asInstanceOf[PeriodicSeriesPlan].stepMs
+      // Time at which raw data would be retrieved which is used to get partition assignments.
+      // It should have time with offset and lookback as we need raw data at time including offset and lookback.
+      val queryTimeRange = TimeRange(periodicSeriesTimeWithOffset.startMs - lookBackMs,
+        periodicSeriesTimeWithOffset.endMs)
+
+      val partitions = partitionLocationProvider.getPartitions(routingKeyMap, queryTimeRange).
+        sortBy(_.timeRange.startMs)
+      var prevPartitionStart = periodicSeriesTimeWithOffset.startMs
+      val execPlans = partitions.zipWithIndex.map { case (p, i) =>
+        // First partition should start from query start time
+        val startMs = if (i == 0) periodicSeriesTime.startMs
```

Comment:
Instead of modifying the timerange for each partition assignments. Is it possible to get unique list of partitions where the routingKey was present for the original time range and fire queries to those unique partition with the original time range only? Later stitch the results back. Eg: If a routing key for time range 1am to 10pm are present in p001 for 1am to 4am, p002 for 4am to 1pm and again in p001 until 10pm. In that case, instead of splitting the timerange and creating 3 queries, just create 2 queries against p001 and p002 for the original time range 1am to 10pm only.

Reply:
As mentioned offline, it may be simpler to break into smaller time range queries anyway. Lets optimize when really needed.